### PR TITLE
Add feedback store

### PR DESCRIPTION
**Pull request by mcbouslog**
Package: lib-classifier

Towards #205, after #531.

Describe your changes:
- add basic feedback store

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?
- [ ] Is the changelog updated?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?

### CODE REVIEW

File: packages/lib-classifier/src/store/FeedbackStore.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import { autorun } from 'mobx'
+import { addDisposer, getRoot, types } from 'mobx-state-tree'
+import _ from 'lodash'
```
Comment (rogerhutchings) [inline packages/lib-classifier/src/store/FeedbackStore.js]: `import clone from 'lodash/clone'` for tree shaking

File: packages/lib-classifier/src/store/FeedbackStore.spec.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import FeedbackStore from './FeedbackStore'
+// import { ProjectFactory, SubjectFactory, WorkflowFactory, SingleChoiceTaskFactory } from '../../test/factories'
```
Comment (rogerhutchings) [inline packages/lib-classifier/src/store/FeedbackStore.spec.js]: Let's trash this if it's not needed

File: packages/lib-classifier/src/store/ClassificationStore.js [hunk 3]
```diff
@@ -118,6 +118,8 @@ const ClassificationStore = types
       if (classification) {
         const annotation = classification.annotations.get(task.taskKey) || createDefaultAnnotation(task)
         annotation.value = annotationValue
+        const feedback = getRoot(self).feedback
+        if (feedback.isActive) feedback.update(annotation)
```
Comment (rogerhutchings) [inline packages/lib-classifier/src/store/ClassificationStore.js]: I'm not a fan of `if`s without blocks

File: packages/lib-classifier/src/store/FeedbackStore.js (hunk 1 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/FeedbackStore.js]: Where MST has methods for acting on models, should we prefer those over lodash in the feedback code? Not that I know if MST has methods to clone objects.

File: packages/lib-classifier/src/store/ClassificationStore.js (hunk 3 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/ClassificationStore.js]: I'd drop this completely and expect `feedback.update()` to make any checks like this.

Comment (eatyourgreens) [reply packages/lib-classifier/src/store/ClassificationStore.js]: Alternatively, wire up `onAction` to fire `feedback.update` on `createDefaultAnnotation`..

File: packages/lib-classifier/src/store/FeedbackStore.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+const FeedbackStore = types
+  .model('FeedbackStore', {
+    isActive: types.optional(types.boolean, false),
+    rules: types.frozen({})
+  })
+  .actions(self => {
+    function afterAttach () {
+      createSubjectObserver()
+    }
+
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/store/FeedbackStore.js]: Maybe wire up an `onAction` observer here (as per previous comment) then fire the appropriate feedback actions in response to classifier state changes.

File: packages/lib-classifier/src/store/FeedbackStore.spec.js (hunk 2 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/FeedbackStore.spec.js]: If we're testing feedback, then we'll need the subject factory, and also a mock classification? Maybe a workflow too.

**Comment by eatyourgreens** [comment]
I've rebased this to bring in changes from the base branch.

File: packages/lib-classifier/src/store/FeedbackStore.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+      self.isActive = helpers.isFeedbackActive(project, subject, workflow)
+
+      if (self.isActive) {
+        self.rules = helpers.generateRules(subject, workflow)
+      }
+    }
+
+    function update (annotation) {
+      const newRules = _.clone(self.rules)
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/store/FeedbackStore.js]: `getSnapshot(self.rules)`?

File: packages/lib-classifier/src/components/Classifier/components/Feedback/components/feedback-modal.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+import Translate from 'react-translate-component';
+import counterpart from 'counterpart';
+import FrameViewer from '../../../../components/frame-viewer';
+import ModalFocus from '../../../../components/modal-focus';
+
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/components/Classifier/components/Feedback/components/feedback-modal.js]: Most of these imports aren't used in the new classifier, so we're better off just building a new Feedback component from scratch.

File: packages/lib-classifier/src/store/FeedbackStore.js (hunk 4 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/FeedbackStore.js]: I'm inclined to just delete 2d81f7177117d7942ccbe7a22d98365fb19791a7.

File: packages/lib-classifier/src/components/Classifier/components/Feedback/components/feedback-modal.js (hunk 6 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/Feedback/components/feedback-modal.js]: I'm inclined to just delete 2d81f7177117d7942ccbe7a22d98365fb19791a7. I don't think we're going to use any of the files in that commit.

**Comment by eatyourgreens** [comment]
http://localhost.zooniverse.org:3000/projects/markb-panoptes/feedback-testing-mb/classify?env=staging to test this locally.

File: packages/lib-classifier/src/store/ClassificationStore.js [hunk 7]
```diff
@@ -42,6 +42,24 @@ const ClassificationStore = types
   .actions(self => {
     function afterAttach () {
       createSubjectObserver()
+
+      onAction(self, (call) => {
+        const feedback = getRoot(self).feedback
+        switch (call.name) {
+          case 'addAnnotation': {
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/store/ClassificationStore.js]: We don't really need to compute feedback each time an annotation changes. We only really need to complete feedback when a classification is complete, at which point we can iterate through the annotations and call `feedback.update` on each one.

**Comment by eatyourgreens** [comment]
Feedback is being saved to the classification as an observable map, so I'm seeing it being POST'ed as {} when I try this branch out on staging.

**Comment by srallen** [comment]
`toJS` can be used to convert the map prior to POSTing and it has a setting for how to convert it. The default is to convert to an object: https://mobx.js.org/refguide/tojson.html

**Comment by eatyourgreens** [comment]
Thanks, but the code's currently using `metadata.feedback = toJS(feedback.rules)`. Even with converting to an object turned off, the output is an empty object.

Looking further down the code, the answer seems to be to use `convertMapToArray()`.
`classificationToSubmit.annotations = convertMapToArray(classificationToSubmit.annotations)`.

**Comment by srallen** [comment]
Oh, yes, by that point, I think `toJS` was already called. `convertMapToArray` will output an array of pairs, so if that's the format needed, then that can be used.

**Comment by eatyourgreens** [comment]
This should be more or less working now, but it's quite messy. Tests are failing because I changed the `addAnnotation` action. It might be better to replace the current code with something that runs on classification complete, iterates through the completed annotations and computes feedback for each one.

File: packages/lib-classifier/src/store/FeedbackStore.js (hunk 1 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/FeedbackStore.js]: I got rid of this by using an observable map for feedback rules, rather than a frozen object.

**Comment by eatyourgreens** [comment]
Closing in favour of #540.

**Pull request closed without merging**